Index: nginx/src/http/modules/ngx_http_proxy_module.c
===================================================================
--- nginx.orig/src/http/modules/ngx_http_proxy_module.c	2013-10-06 07:41:52.758820066 +0900
+++ nginx/src/http/modules/ngx_http_proxy_module.c	2013-10-06 07:41:52.754820182 +0900
@@ -2796,8 +2796,10 @@
 
     if (conf->upstream.upstream || conf->proxy_lengths) {
         clcf = ngx_http_conf_get_module_loc_conf(cf, ngx_http_core_module);
-        if (clcf->handler == NULL && clcf->lmt_excpt) {
-            clcf->handler = ngx_http_proxy_handler;
+        if (clcf->handler == NULL) {
+            if (clcf->lmt_excpt) {
+                clcf->handler = ngx_http_proxy_handler;
+            }
             conf->location = prev->location;
         }
     }
