--- a/src/core/nginx.h
+++ b/src/core/nginx.h
@@ -11,7 +11,11 @@
 
 #define nginx_version      1004001
 #define NGINX_VERSION      "1.4.1"
+#ifdef NGINX_OPENRESTY_VERSION
+#define NGINX_VER          "ngx_openresty/" NGINX_VERSION ".1"
+#else
 #define NGINX_VER          "nginx/" NGINX_VERSION
+#endif
 
 #define NGINX_VAR          "NGINX"
 #define NGX_OLDPID_EXT     ".oldbin"
--- a/src/http/ngx_http_header_filter_module.c
+++ b/src/http/ngx_http_header_filter_module.c
@@ -46,7 +46,11 @@
 };
 
 
+#ifdef NGINX_OPENRESTY_VERSION
+static char ngx_http_server_string[] = "Server: ngx_openresty" CRLF;
+#else
 static char ngx_http_server_string[] = "Server: nginx" CRLF;
+#endif
 static char ngx_http_server_full_string[] = "Server: " NGINX_VER CRLF;
 
 
