Index: nginx/src/os/unix/ngx_channel.c
===================================================================
--- nginx.orig/src/os/unix/ngx_channel.c	2013-10-06 07:41:41.211154148 +0900
+++ nginx/src/os/unix/ngx_channel.c	2013-10-06 07:41:41.203154380 +0900
@@ -31,6 +31,8 @@
         msg.msg_controllen = 0;
 
     } else {
+        ngx_memzero(&cmsg, sizeof(cmsg));
+
         msg.msg_control = (caddr_t) &cmsg;
         msg.msg_controllen = sizeof(cmsg);
 
Index: nginx/src/os/unix/ngx_process_cycle.c
===================================================================
--- nginx.orig/src/os/unix/ngx_process_cycle.c	2013-10-06 07:41:41.211154148 +0900
+++ nginx/src/os/unix/ngx_process_cycle.c	2013-10-06 07:41:41.207154264 +0900
@@ -355,6 +355,8 @@
 
     ngx_log_error(NGX_LOG_NOTICE, cycle->log, 0, "start worker processes");
 
+    ngx_memzero(&ch, sizeof(ngx_channel_t));
+
     ch.command = NGX_CMD_OPEN_CHANNEL;
 
     for (i = 0; i < n; i++) {
@@ -460,6 +462,8 @@
     ngx_err_t      err;
     ngx_channel_t  ch;
 
+    ngx_memzero(&ch, sizeof(ngx_channel_t));
+
 #if (NGX_BROKEN_SCM_RIGHTS)
 
     ch.command = 0;
@@ -561,6 +565,8 @@
     ngx_channel_t     ch;
     ngx_core_conf_t  *ccf;
 
+    ngx_memzero(&ch, sizeof(ngx_channel_t));
+
     ch.command = NGX_CMD_CLOSE_CHANNEL;
     ch.fd = -1;
 
